src/Features/Core/Portable/Navigation/INavigableDocumentLocation.cs

```diff
@@ -0,0 +1,38 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Navigation
+{
+    internal interface INavigableDocumentLocation
+    {
+        /// <summary>
+        /// Navigates to a location in a particular document, opening it if necessary.  This work must happen quickly.
+        /// Any expensive async work must be done in the corresponding IDocumentNavigationService.GetNavigableLocation
+        /// method.  This method is async only to allow final clients to call this from a non-UI thread while allowing
+        /// the navigation to jump to the UI thread.
+        /// </summary>
+        Task<bool> NavigateToAsync(CancellationToken cancelletionToken);
```

Comment:
note: for this PR/set-of-refactorings, i'd like to have this async for the moment.  then, once all the work is done, we should have a good idea if it can be safely made sync.

Reply:
I'm OK with this, but let's keep an eye out for places the "this should be async only..." rule gets violated. And maybe we will still have cases where this needs to be async; I could imagine in VS even opening a file this may still have some async bits since we at least have to check if the file exists, or something.